python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -1704,74 +1625,267 @@ def can_deserialize(
+            isinstance(other, TimePartitionKeyPartitionsSubset)
+            and self.partitions_def == other.partitions_def
+            and self._included_partition_keys == other._included_partition_keys
+        ) or super(TimePartitionKeyPartitionsSubset, self).__eq__(other)
+
     @classmethod
     def empty_subset(cls, partitions_def: PartitionsDefinition) -> "PartitionsSubset":
         if not isinstance(partitions_def, TimeWindowPartitionsDefinition):
             check.failed("Partitions definition must be a TimeWindowPartitionsDefinition")
         partitions_def = cast(TimeWindowPartitionsDefinition, partitions_def)
-        return cls(partitions_def, 0, [], set())
-
-    def serialize(self) -> str:
-        return json.dumps(
-            {
-                "version": self.SERIALIZATION_VERSION,
-                # included_time_windows is already sorted, so no need to sort here to guarantee
-                # stable serialization between identical subsets
-                "time_windows": [
-                    (window.start.timestamp(), window.end.timestamp())
-                    for window in self.included_time_windows
-                ],
-                "num_partitions": self._num_partitions,
-            }
-        )
+        return cls(partitions_def, set())
 
     def with_partitions_def(
         self, partitions_def: TimeWindowPartitionsDefinition
-    ) -> "TimeWindowPartitionsSubset":
+    ) -> "BaseTimeWindowPartitionsSubset":
         check.invariant(
             partitions_def.cron_schedule == self._partitions_def.cron_schedule,
             "num_partitions would become inaccurate if the partitions_defs had different cron"
             " schedules",
         )
-        return TimeWindowPartitionsSubset(
+        return TimePartitionKeyPartitionsSubset(
             partitions_def=partitions_def,
-            num_partitions=self.num_partitions,
-            included_time_windows=self._included_time_windows,
             included_partition_keys=self._included_partition_keys,
         )
 
+    def resolve(self) -> "TimeWindowPartitionsSubset":
+        return TimeWindowPartitionsSubset(
+            partitions_def=self._partitions_def,
+            num_partitions=self.num_partitions,
+            included_time_windows=self.included_time_windows,
+        )
+
+
+class TimeWindowPartitionsSubset(BaseTimeWindowPartitionsSubset):
+    def __init__(
+        self,
+        partitions_def: TimeWindowPartitionsDefinition,
+        num_partitions: Optional[int] = None,
+        included_time_windows: Sequence[TimeWindow] = [],
+    ):
+        check.opt_int_param(num_partitions, "num_partitions")
+        self._partitions_def = check.inst_param(
+            partitions_def, "partitions_def", TimeWindowPartitionsDefinition
+        )
+        self._num_partitions = (
+            num_partitions
+            if num_partitions
+            else self._num_partitions_from_time_windows(partitions_def, included_time_windows)
```

Comment:
This can be an expensive operation. If it wasn't previously part of the constructor, I'd be nervous that adding it could cause perf issues.

Reply:
got it -- I amended this to instead just be optional, and to only calculate on demand